### PR TITLE
Settle loan installments from bank transactions

### PR DESCRIPTION
**Pull request by patrickweh**
Loan installments were settled by a manual `is_paid` flag. A repayment that never arrived, or one that was returned, stayed invisible in the schedule: the plan only ever showed what the remaining balance ought to be.

## What changes

Installments and transactions are linked through a new `loan_installment_transaction` pivot. The accepted assignments of an installment decide whether it is settled, so a returned direct debit cancels its payment out and reopens the installment, because it arrives as a positive counter booking on the same installment. `remaining`, `total_interest` and `progress` on the loan are recalculated from there and stay sortable and summable in data tables.

The loan detail page gains a payments tab listing every booking with date, installment number, note and amount, each linked to its transaction. The transaction assignment screen can assign, edit and detach loan installments, counts loan suggestions in its tab and accepts them along with order suggestions.

## Matching

`MatchTransactionsWithOrderJob` shrinks to a dispatcher. The order logic moved verbatim into `OrderMatcher`; `LoanInstallmentMatcher` runs first, because both matchers consume the same transaction balance and the loan signals are the harder ones.

A transaction matches through the creditor iban or the loan number in its purpose. It is only assigned when the amount equals the installment amount or the purpose names the loan, so an unrelated booking from the same bank is left alone. A positive amount is treated as a chargeback and searches covered installments instead of open ones. Auto acceptance stays gated on `auto_accept_secure_transaction_matches`, an exact amount and a booking date within ten days of the due date.

## Tenant scope

`Loan` used `HasTenantAssignment` but never declared a `tenant` relation, so `UserTenantScope` had nothing to filter on and loans of every tenant were visible. The relation is added, and installments inherit the scope through their loan.

## Also

`SearchController` now resolves dotted search fields as relations, which is what the installment select needs to search by loan name and number.

## Summary by Sourcery

Link loan installments to bank transactions, drive loan progress from these assignments, and extend transaction matching and UI to support loan repayments alongside orders.

New Features:
- Introduce pivot model, migration, actions, and Livewire form to assign transactions to loan installments and manage their lifecycle.
- Add a payments tab and overdue/settlement status tracking to loan detail views, with loan list showing overdue installment counts.
- Enable manual assignment and editing of loan-installment links in the transaction assignment UI, including bulk acceptance of loan suggestions.
- Add dedicated matchers for loan installments and orders so automatic transaction matching prioritizes loan repayments over orders based on IBAN, loan number, and amount.

Bug Fixes:
- Scope loans and installments by tenant so users only see loan data belonging to their tenants.
- Include loan-installment assignments when recalculating transaction balances, ensuring balances reflect accepted loan repayments.

Enhancements:
- Refactor order matching logic from the transaction matching job into a reusable OrderMatcher service.
- Update loan totals and remaining balance calculations to derive settlement state from covered installments rather than a simple is_paid flag.
- Allow SearchController to interpret dotted search fields as relation paths for server-side searching.

Tests:
- Add extensive feature and Livewire tests for loan-installment matching, tenant scoping, loan coverage edge cases, transaction assignments, and search relation handling.

### CODE REVIEW

File: src/Actions/LoanInstallmentTransaction/UpdateLoanInstallmentTransaction.php [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+    {
+        return UpdateLoanInstallmentTransactionRuleset::class;
+    }
+
+    public function performAction(): LoanInstallmentTransaction
+    {
+        $loanInstallmentTransaction = resolve_static(LoanInstallmentTransaction::class, 'query')
+            ->whereKey($this->getData('pivot_id'))
+            ->first();
+        $loanInstallmentTransaction->fill($this->getData());
```
Comment (SirSplasch) [inline src/Actions/LoanInstallmentTransaction/UpdateLoanInstallmentTransaction.php:27]: new line before

File: src/Rulesets/LoanInstallmentTransaction/CreateLoanInstallmentTransactionRuleset.php [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+                app(ModelExists::class, ['model' => LoanInstallment::class]),
+            ],
+            'transaction_id' => [
+                'required',
+                'integer',
+                app(ModelExists::class, ['model' => Transaction::class]),
+            ],
+            'amount' => [
+                'required',
+                app(Numeric::class),
```
Comment (SirSplasch) [inline src/Rulesets/LoanInstallmentTransaction/CreateLoanInstallmentTransactionRuleset.php:28]: Does this have a minimum, like can the amount be negative?

File: src/Rulesets/LoanInstallmentTransaction/UpdateLoanInstallmentTransactionRuleset.php [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+            'pivot_id' => [
+                'required',
+                'integer',
+                app(ModelExists::class, ['model' => LoanInstallmentTransaction::class]),
+            ],
+            'loan_installment_id' => [
+                'sometimes',
+                'required',
+                'integer',
+                app(ModelExists::class, ['model' => LoanInstallment::class]),
```
Comment (SirSplasch) [inline src/Rulesets/LoanInstallmentTransaction/UpdateLoanInstallmentTransactionRuleset.php]: add subject + subjectKeyName to ModelExists

File: src/Rulesets/LoanInstallmentTransaction/UpdateLoanInstallmentTransactionRuleset.php [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+            ],
+            'transaction_id' => [
+                'sometimes',
+                'required',
+                'integer',
+                app(ModelExists::class, ['model' => Transaction::class]),
```
Comment (SirSplasch) [inline src/Rulesets/LoanInstallmentTransaction/UpdateLoanInstallmentTransactionRuleset.php]: add subject + subjectKeyName to ModelExists

File: src/Rulesets/LoanInstallmentTransaction/UpdateLoanInstallmentTransactionRuleset.php [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+            ],
+            'amount' => [
+                'sometimes',
+                'required',
+                app(Numeric::class),
```
Comment (SirSplasch) [inline src/Rulesets/LoanInstallmentTransaction/UpdateLoanInstallmentTransactionRuleset.php:45]: Can this be negative?

File: src/Models/Pivots/LoanInstallmentTransaction.php [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+        $previousInstallmentId = $this->getOriginal('loan_installment_id');
+
+        if (! is_null($previousInstallmentId) && $previousInstallmentId !== $this->loan_installment_id) {
+            resolve_static(LoanInstallment::class, 'query')
+                ->whereKey($previousInstallmentId)
+                ->first()
+                ?->recalculateLoan();
+        }
+
+        $this->loanInstallment()->first()?->recalculateLoan();
```
Comment (SirSplasch) [inline src/Models/Pivots/LoanInstallmentTransaction.php]: wrap
->first()
?->recalculateLoan();

File: src/Models/Pivots/LoanInstallmentTransaction.php [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+
+        if ($this->is_accepted || $this->wasChanged('is_accepted') || ! $this->exists) {
+            $this->transaction()->first()?->calculateBalance()->save();
```
Comment (SirSplasch) [inline src/Models/Pivots/LoanInstallmentTransaction.php]: wrap
->first()
?->calculateBalance()
->save();

File: src/Models/Pivots/LoanInstallmentTransaction.php [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+namespace FluxErp\Models\Pivots;
+
+use FluxErp\Models\LoanInstallment;
+use FluxErp\Models\Transaction;
+use FluxErp\Traits\Model\HasPackageFactory;
+use Illuminate\Database\Eloquent\Relations\BelongsTo;
+
+class LoanInstallmentTransaction extends FluxPivot
+{
+    use HasPackageFactory;
```
Comment (SirSplasch) [inline src/Models/Pivots/LoanInstallmentTransaction.php:12]: the factory for this pivot model is missing.

File: src/Models/LoanInstallment.php [hunk 9]
```diff
@@ -58,13 +71,72 @@ public function loan(): BelongsTo
...
-        $this->loan
+        $this->loan()
+            ->first()
             ?->calculateRemaining()
             ->calculateTotalInterest()
             ->calculateProgress()
             ->save();
     }
+
+    // Scopes
```
Comment (SirSplasch) [inline src/Models/LoanInstallment.php:115]: make the scope methods protected

File: src/Models/LoanInstallment.php [hunk 10]
```diff
@@ -58,13 +71,72 @@ public function loan(): BelongsTo
...
+    }
+
+    public function getDescription(): ?string
+    {
+        return trans('Due Date') . ' ' . $this->due_date?->locale(app()->getLocale())->isoFormat('L');
+    }
+
+    public function getLabel(): ?string
+    {
+        return trim(($this->loan?->name ?? '') . ' ' . trans('Sequence') . ' ' . $this->sequence);
```
Comment (SirSplasch) [inline src/Models/LoanInstallment.php]: __() instead of trans()

File: src/Models/LoanInstallment.php [hunk 11]
```diff
@@ -58,13 +71,72 @@ public function loan(): BelongsTo
...
+
     // Public methods
+    public function getAvatarUrl(): ?string
+    {
+        return null;
+    }
+
+    public function getDescription(): ?string
+    {
+        return trans('Due Date') . ' ' . $this->due_date?->locale(app()->getLocale())->isoFormat('L');
```
Comment (SirSplasch) [inline src/Models/LoanInstallment.php]: __() instead of trans()

File: src/Models/LoanInstallment.php [hunk 12]
```diff
@@ -27,6 +27,11 @@ class LoanInstallment extends FluxModel
 
     protected static function booted(): void
     {
+        static::addGlobalScope(
+            'tenant',
```
Comment (SirSplasch) [inline src/Models/LoanInstallment.php]: Why is this scope called 'tenant' although it only checks whereHas('loan')?
Also why define this scope like that and not by creating an actual scope class?

File: src/Models/Transaction.php [hunk 13]
```diff
@@ -73,6 +74,18 @@ public function ledgerAccountTransactions(): HasMany
         return $this->hasMany(LedgerAccountTransaction::class);
     }
 
+    public function loanInstallmentTransactions(): HasMany
+    {
+        return $this->hasMany(LoanInstallmentTransaction::class);
+    }
+
+    public function loanInstallments(): BelongsToMany
```
Comment (SirSplasch) [inline src/Models/Transaction.php:77]: loanInstallments above loanInstallmentTransactions

File: src/Livewire/Accounting/Loan.php [hunk 14]
```diff
@@ -159,51 +170,127 @@ public function scheduleHeaders(): array
...
+            ->reduce(
+                fn (string $carry, TransactionModel $transaction) => bcadd(
+                    $carry,
+                    (string) $transaction->pivot->amount,
+                    2
+                ),
+                '0'
+            );
+
+        return bccomp($sum, '0', 2) === -1 ? bcabs($sum) : '0.00';
```
Comment (SirSplasch) [inline src/Livewire/Accounting/Loan.php:250]: Just to be sure, if $sum is positive and above 0, it will be 0.00, is this correct?

File: src/Livewire/Accounting/Loan.php [hunk 15]
```diff
@@ -159,51 +170,127 @@ public function scheduleHeaders(): array
...
+        $interest = '0';
+        $paidPrincipal = '0';
+        $paidInterest = '0';
+
+        foreach ($this->loadInstallments($loan) as $installment) {
+            $principal = bcadd($principal, (string) $installment->principal_amount, 2);
+            $interest = bcadd($interest, (string) $installment->interest_amount, 2);
+
+            if ($installment->is_paid
+                || bccomp($this->coveredAmount($installment), $installment->getTotalAmount(), 2) > -1
```
Comment (SirSplasch) [inline src/Livewire/Accounting/Loan.php]: !== -1

**Review by SirSplasch — CHANGES_REQUESTED** [review]

File: src/Rulesets/LoanInstallmentTransaction/CreateLoanInstallmentTransactionRuleset.php (hunk 2 above)
Comment (patrickweh) [reply src/Rulesets/LoanInstallmentTransaction/CreateLoanInstallmentTransactionRuleset.php:28]: Negative is the normal case here. A repayment leaves the account, so it is assigned as a negative amount, and a returned direct debit comes back positive on the same installment. Coverage is the sum of both, which is how a return cancels its payment out and reopens the installment. A minimum would break exactly that.

File: src/Rulesets/LoanInstallmentTransaction/UpdateLoanInstallmentTransactionRuleset.php (hunk 5 above)
Comment (patrickweh) [reply src/Rulesets/LoanInstallmentTransaction/UpdateLoanInstallmentTransactionRuleset.php:45]: Yes, same as on create. Repayments are negative, returned direct debits positive, and the sum of both decides whether the installment is settled.

File: database/factories/LoanInstallmentTransactionFactory.php [hunk 16]
```diff
@@ -0,0 +1,19 @@
...
+
+class LoanInstallmentTransactionFactory extends Factory
+{
+    protected $model = LoanInstallmentTransaction::class;
+
+    public function definition(): array
+    {
+        return [
+            'amount' => fake()->randomFloat(2, -10000, 0),
+            'is_accepted' => fake()->boolean(),
```
Comment (SirSplasch) [inline database/factories/LoanInstallmentTransactionFactory.php:17]: can add note key

**Review by SirSplasch — CHANGES_REQUESTED** [review]

**Review by SirSplasch — APPROVED** [review]
